web/modules/custom/employees/src/Plugin/Action/AddGroupMembershipsAction.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Drupal\employees\Plugin\Action;
+
+use Drupal\views_bulk_operations\Action\ViewsBulkOperationsActionBase;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Action\Attribute\Action;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\PluginFormInterface;
+use Drupal\group\Entity\Group;
+use Drupal\Core\Access\AccessResult;
+
+/**
+ * Views bulk operations action that allows adding users to multiple groups in bulk.
+ *
+ * @Action(
+ *   id = "employees_add_user_to_groups",
+ *   label = @Translation("Add or update members (custom action)"),
+ *   type = "group",
+ *   confirm = FALSE,
+ * )
+ */
+final class AddGroupMembershipsAction extends ViewsBulkOperationsActionBase implements PluginFormInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute(Group $group = NULL) {
+    if ($group === NULL || \count($this->configuration['user_id']) === 0) {
+      return $this->t('Invalid entity or configuration.');
+    }
+
+    $user_ids = [];
+    foreach ($this->configuration['user_id'] as $item) {
+      $user_ids[$item['target_id']] = $item['target_id'];
+    }
+
+    $role_ids = $this->configuration['role_ids'];
+    foreach (\Drupal::entityTypeManager()->getStorage('user')->loadMultiple($user_ids) as $user) {
+      $membership = $group->getMember($user);
+      if ($membership === FALSE) {
+        // Add user to the group with the assigned role(s)
+        $role_ids = array_diff($role_ids, [0]);   // Remove array elements with the value of "0"
+        $group->addMember($user, ['group_roles' => array_keys($role_ids)]);
+      } else {
+        // User is already a member of the group so just update the role(s)
+        foreach ($role_ids as $role_id => $assign_role) {
+          if ($assign_role) {
+            $membership->addRole($role_id);
+          } else {
+            $membership->removeRole($role_id);
+          }
+        }
+      }
+    }
+
+    return $this->t('User has been added to the group with the assigned roles.');
+  }
+
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form['user_id'] = [
+      '#type' => 'entity_autocomplete',
+      '#title' => $this->t('User'),
+      '#placeholder' => $this->t('Enter user name'),
+      '#target_type' => 'user',
+      '#selection_settings' => ['filter' => ['access' => 'administer members']],
+      '#required' => TRUE,
+      '#tags' => TRUE,
+      '#description' => $this->t('Use a comma to select multiple users.'),
+    ];
+    
+    // Get the group roles associated with the first selected group
+    // HACK ALERT: This code assumes that all selected groups are of the same group type
```

Comment:
I think this is a non-issue on the public site, since it's probably rare that someone would be modifying a City Official group membership in bulk along with a Base Group, but I could see it being an issue on Employees with the proliferation of two different open/restricted group types.

If it's not easy to support both group types, could we add some type of warning that all group types need to be the same? As it stands, you get an "unexpected error" message if trying to bulk add/update membership on two different group types.

If neither is feasible, I suppose we can go ahead and merge. Everything else looks good!

Reply:
Good catch. I was able to make it display an error and stop the user if they select multiple group types.